PERF: Add cache keyword to to_datetime (#11665)

mroeschke opened the pull request:
- [x] closes #11665
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [x] whatsnew entry

Added a `cache` keyword to `to_datetime` to speedup parsing of duplicate dates:

Some notes:

1) I defaulted `cache=False` i.e. don't use a cache to parse the dates. Should the default be `True`? 

2) I used `pd.unique()` to identify unique dates, and the current implementation did not accept a tuple of strings (objects). I added `tuple_to_object_array` and patched ` _ensure_arraylike` to fix this.

3) There is currently an included test that fails due to a case when using `to_datetime(..., utc=True)` with a `Series`. I am inclined to believe `In[5]` should have been the existing behavior. Thoughts?

```
In [2]: test_dates = ['20130101 00:00:00'] * 10

In [3]: s = pd.Series(test_dates)

# Same as existing behavior
In [4]: pd.to_datetime(s, utc=True, cache_datetime=False)
Out[4]: 
0   2013-01-01
1   2013-01-01
2   2013-01-01
3   2013-01-01
4   2013-01-01
5   2013-01-01
6   2013-01-01
7   2013-01-01
8   2013-01-01
9   2013-01-01
dtype: datetime64[ns]

In [5]: pd.to_datetime(s, utc=True, cache_datetime=True)
Out[5]: 
0   2013-01-01 00:00:00+00:00
1   2013-01-01 00:00:00+00:00
2   2013-01-01 00:00:00+00:00
3   2013-01-01 00:00:00+00:00
4   2013-01-01 00:00:00+00:00
5   2013-01-01 00:00:00+00:00
6   2013-01-01 00:00:00+00:00
7   2013-01-01 00:00:00+00:00
8   2013-01-01 00:00:00+00:00
9   2013-01-01 00:00:00+00:00
dtype: datetime64[ns, UTC]

```

gfyoung commented:
Let me address 1 and 3 first: IIUC, this caching should only speed up performance.  The results you get with or without caching should not be different.  The fact that you are getting different results indicates something is going weird with the caching.

gfyoung commented:
For 2, I think that's worthy enough as its own PR.  I'm not sure I fully agree with the patch (i.e. why do we need another special-case `to_object_array` method), but even so, you should separate that out as another PR and submit that for review (along with tests).

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -375,6 +375,23 @@ cpdef ndarray[object] list_to_object_array(list obj):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
+cpdef ndarray[object] tuple_to_object_array(tuple obj):
```
don't do this, you are dupliating lots of code. I suspect you are hitting

```
In [2]: pd.unique(('foo'),)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-bd2fadf86f43> in <module>()
----> 1 pd.unique(('foo'),)

/Users/jreback/pandas/pandas/core/algorithms.py in unique(values)
    348     """
    349 
--> 350     values = _ensure_arraylike(values)
    351 
    352     # categorical is a fast-path

/Users/jreback/pandas/pandas/core/algorithms.py in _ensure_arraylike(values)
    171         inferred = lib.infer_dtype(values)
    172         if inferred in ['mixed', 'string', 'unicode']:
--> 173             values = lib.list_to_object_array(values)
    174         else:
    175             values = np.asarray(values)

TypeError: Argument 'obj' has incorrect type (expected list, got str)

```

if so, pls do a separate PR for this (it won't involve *any* cython code, just a simple change).

jreback commented on pandas/core/algorithms.py:
```diff
@@ -170,7 +170,10 @@ def _ensure_arraylike(values):
                                ABCIndexClass, ABCSeries)):
         inferred = lib.infer_dtype(values)
         if inferred in ['mixed', 'string', 'unicode']:
-            values = lib.list_to_object_array(values)
```
see above

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -183,7 +184,8 @@ def _guess_datetime_format_for_array(arr, **kwargs):
                 utc=None, box=True, format=None, exact=True,
-                unit=None, infer_datetime_format=False, origin='unix'):
+                unit=None, infer_datetime_format=False, origin='unix',
+                cache_datetime=False):
```
``use_cache=True``, change name and default to True

chris-b1 replied:
Why not call this `cache=True`?

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -257,6 +259,10 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
 
+    cache_datetime : boolean, default False
+        If True, use a cache of unique, converted dates to apply the datetime
+        conversion. Produces signficant speed-ups when parsing duplicate dates
```
see above, add versionadded tag

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -340,6 +346,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
 
     tz = 'utc' if utc else None
 
+    cache = None
```
I *think* can be moved inside ``_convert_list_like`` (maybe)

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -340,6 +346,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
 
+    cache = None
+    if (cache_datetime and is_list_like(arg) and
+       not isinstance(arg, DatetimeIndex)):
```
when you write an asv (see below), this needs a min len arg (maybe 1000)

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -340,6 +346,19 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        # No need to convert with a cache if the arg is already a DatetimeIndex
+        unique_dates = pd.unique(arg)
+        if len(unique_dates) != len(arg):
+            cache = {d: pd.to_datetime(d, errors=errors, dayfirst=dayfirst,
```
this is very inefficient, you are converting element by-element, simply

```
Series(to_datetime(unique_dates.....), index=unique_dates)
```

also avoids iterating over things

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -505,15 +524,27 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+            result = arg.map(cache)
+        else:
+            values = _convert_listlike(arg._values, False, format)
+            result = pd.Series(values, index=arg.index, name=arg.name)
```
leave the imports alone

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -1,5 +1,6 @@
 from datetime import datetime, timedelta, time
 import numpy as np
+import pandas as pd
```
use ``algorithms.unique`` below

jreback commented on pandas/tests/indexes/datetimes/test_tools.py:
```diff
@@ -306,6 +306,45 @@ def test_to_datetime_tz_psycopg2(self):
                                     dtype='datetime64[ns, UTC]')
         tm.assert_index_equal(result, expected)
 
+    @pytest.mark.parametrize("box", [True, False])
```
this test is ok, but rather I would like to see a parametrize on pretty much every other test function in this file. that tests both ``use_cache=True/False``

jreback reviewed (changes requested):
need asv!

bench the cartesian product of several dtypes (int with unit, str, str with format, dti) and sizes (1000, 100000)
 and all dupes, uniques

I suspect under a certain size cache doesn't matter/help, need to see where that is.

mroeschke commented:
@gfyoung In regards to my 3rd point, I agree that the results should be the same whether `cache_datetime=True/False`. I am just curious whether `pd.to_datetime(Series(...), utc=True)` should return a `dtype: datetime64[ns, UTC]` in the first place.

```
In [9]: pd.__version__
Out[9]: u'0.20.3'

In [10]: test_dates = ['20130101 00:00:00'] * 10

In [11]: s = pd.Series(test_dates)

# Should this result have a datetime64[ns, UTC] dtype like Out [13]?
In [12]: pd.to_datetime(s, utc=True)
Out[12]:
0   2013-01-01
1   2013-01-01
2   2013-01-01
3   2013-01-01
4   2013-01-01
5   2013-01-01
6   2013-01-01
7   2013-01-01
8   2013-01-01
9   2013-01-01
dtype: datetime64[ns]

In [13]: pd.to_datetime(test_dates, utc=True)
Out[13]:
DatetimeIndex(['2013-01-01', '2013-01-01', '2013-01-01', '2013-01-01',
               '2013-01-01', '2013-01-01', '2013-01-01', '2013-01-01',
               '2013-01-01', '2013-01-01'],
              dtype='datetime64[ns, UTC]', freq=None)

# I think this is essentially what my caching implementation does. 
In [14]: pd.Series([pd.Timestamp('20130101 00:00:00', tz='utc')]*10)
Out[14]:
0   2013-01-01 00:00:00+00:00
1   2013-01-01 00:00:00+00:00
2   2013-01-01 00:00:00+00:00
3   2013-01-01 00:00:00+00:00
4   2013-01-01 00:00:00+00:00
5   2013-01-01 00:00:00+00:00
6   2013-01-01 00:00:00+00:00
7   2013-01-01 00:00:00+00:00
8   2013-01-01 00:00:00+00:00
9   2013-01-01 00:00:00+00:00
dtype: datetime64[ns, UTC]
```

This discussion may be more appropriate in a separate issue if it is one. I may not hit this though once I refactor this implementation.

jreback commented:
there is another issue about utc=True

so it's out of scope for this PR

mroeschke commented:
Thanks for confirming @jreback and @gfyoung. I will work on implementing your suggestions.

mroeschke committed:
move cache into convert_listlike

mroeschke committed:
Move cache down the stack, explore threshold to trigger cache

mroeschke committed:
Add more cache conditions

mroeschke committed:
Add some benchmarks

mroeschke committed:
Some performance testing

jreback commented:
can you rebase / update

mroeschke committed:
Add asvs, modify tests for caches

mroeschke committed:
Fix asv errors and condition

pep8speaks commented:
Hello @mroeschke! Thanks for updating the PR.







 - In the file [`pandas/tests/indexes/datetimes/test_tools.py`](https://github.com/pandas-dev/pandas/blob/72e99da0796171307553531292a85aa774879beb/pandas/tests/indexes/datetimes/test_tools.py), following are the PEP8 issues :

> [Line 224:69](https://github.com/pandas-dev/pandas/blob/72e99da0796171307553531292a85aa774879beb/pandas/tests/indexes/datetimes/test_tools.py#L224): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 403:1](https://github.com/pandas-dev/pandas/blob/72e99da0796171307553531292a85aa774879beb/pandas/tests/indexes/datetimes/test_tools.py#L403): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1121:9](https://github.com/pandas-dev/pandas/blob/72e99da0796171307553531292a85aa774879beb/pandas/tests/indexes/datetimes/test_tools.py#L1121): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 1121:9](https://github.com/pandas-dev/pandas/blob/72e99da0796171307553531292a85aa774879beb/pandas/tests/indexes/datetimes/test_tools.py#L1121): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

mroeschke committed:
Pep8 fixes

mroeschke commented:
Here are the results from `timeseries.ToDatetime` ASVs.

```
>>>asv continuous -f 1.1 upstream/master fix_11665 -b timeseries.ToDatetime
      before           after         ratio
     [e0fe5cc6]       [72e99da0]
+     7.91±0.04ms           16.5ms     2.08  timeseries.ToDatetime.time_iso8601_format_no_sep
+     8.07±0.03ms      16.5±0.04ms     2.04  timeseries.ToDatetime.time_iso8601_nosep
+     8.18±0.01ms      16.4±0.03ms     2.00  timeseries.ToDatetime.time_iso8601
+     8.34±0.02ms      16.5±0.01ms     1.98  timeseries.ToDatetime.time_iso8601_format
+     12.5±0.07ms      14.8±0.05ms     1.18  timeseries.ToDatetime.time_format_YYYYMMDD
-        3.49±0ms         2.98±0ms     0.86  timeseries.ToDatetime.time_cache_with_dup_string_dates_and_format
-     3.54±0.01ms      2.98±0.01ms     0.84  timeseries.ToDatetime.time_cache_with_dup_string_dates
-     8.96±0.03ms      7.12±0.01ms     0.79  timeseries.ToDatetime.time_cache_with_dup_seconds_and_unit
-           2.11s       44.1±0.2ms     0.02  timeseries.ToDatetime.time_format_exact
-           2.02s      30.5±0.04ms     0.02  timeseries.ToDatetime.time_format_no_exact
-       407±0.3ms      3.35±0.01ms     0.01  timeseries.ToDatetime.time_cache_with_dup_string_tzoffset_dates

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

mroeschke committed:
Remove unused import

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -259,7 +260,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+        If True, use a cache of unique, converted dates to apply the datetime
+        conversion. Produces signficant speed-ups when parsing duplicate date.
 
+        .. versionadded: 0.20.2
```
0.21.0

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -514,18 +519,41 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        if len(unique_dates) != len(arg):
+            from pandas import Series
+            cache_dates = _convert_listlike(unique_dates, False, format)
+            convert_cache = Series(cache_dates, index=unique_dates)
```
so its better to actually to make ``convert_cache`` a function, which can then take the unique_dates and return the converted data, avoids lots of code duplication.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -334,7 +334,7 @@ def __new__(cls, data=None,
                 is_integer_dtype(data)):
             data = tools.to_datetime(data, dayfirst=dayfirst,
-                                     yearfirst=yearfirst)
+                                     yearfirst=yearfirst, cache=False)
```
reason for this chanege?

mroeschke replied:
This is in order to prevent a `RuntimeError` due to recursion.

The cache is built using `_convert_listlike` which can return a `DatetimeIndex`, then the `DatetimeIndex` constructor can call `to_datetime` which goes back to `_convert_listlike`...

mroeschke committed:
Wrap cache logic in a function

mroeschke committed:
Fix Series test

mroeschke committed:
Add whatsnew and small documentation fix

mroeschke committed:
pep 8 fixes

mroeschke commented:
Tests are passing now and here are the latest asv results:

```
>>> asv continuous -f 1.1 upstream/master fix_11665 -b timeseries.ToDatetime
 before           after         ratio
     [37860a5f]       [bdff633d]
+     8.05±0.05ms      16.3±0.09ms     2.02  timeseries.ToDatetime.time_iso8601
+     7.99±0.06ms      15.8±0.09ms     1.98  timeseries.ToDatetime.time_iso8601_nosep
+     8.53±0.06ms       16.5±0.3ms     1.94  timeseries.ToDatetime.time_iso8601_format_no_sep
+     8.72±0.02ms      16.0±0.04ms     1.84  timeseries.ToDatetime.time_iso8601_format
+      12.5±0.1ms      14.6±0.08ms     1.17  timeseries.ToDatetime.time_format_YYYYMMDD
-        3.53±0ms         3.14±0ms     0.89  timeseries.ToDatetime.time_cache_with_dup_string_dates_and_format
-     8.18±0.03ms         7.25±0ms     0.89  timeseries.ToDatetime.time_cache_with_dup_seconds_and_unit
-           2.07s      47.7±0.02ms     0.02  timeseries.ToDatetime.time_format_exact
-           1.96s      31.7±0.02ms     0.02  timeseries.ToDatetime.time_format_no_exact
-       399±0.5ms      3.51±0.01ms     0.01  timeseries.ToDatetime.time_cache_with_dup_string_tzoffset_dates

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -111,7 +112,11 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
           origin.
 
         .. versionadded: 0.20.0
+    cache : boolean, default False
```
default is True

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -305,6 +310,29 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             except (ValueError, TypeError):
                 raise e
 
+    def _maybe_convert_cache(arg, cache, tz):
```
use a proper doc-string here

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -369,15 +397,29 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        result = _maybe_convert_cache(arg, cache, tz)
+        if result is None:
+            result = _convert_listlike(arg, box, format, name=arg.name)
+        else:
```
why can't you handle these caess list-list/index) *inside* ``_maybe_convert_cache``? (I am talking about the else/box part.

jreback commented:
what is the 2x slowdown on some of the existing tests?

chris-b1 commented:
It looks like the iso8601 path is so fast the caching always hurts - I think that makes sense, that conversion shouldn't be much more expensive than hashing, but not sure how to handle it.

mroeschke committed:
Move box logic into maybe_convert_cache

mroeschke commented:
The 2x slowdown occurred with iso8601 strings dates without any duplicates. I've attached the profile below of the benchmark with the largest slowdown. It looks like it's expensive put the strings (objects) through  `algorithms.unique` and then continue down the regular conversion path since there are no duplicates. 
```
In [3]: rng = date_range(start='1/1/2000', periods=20000, freq='H')

In [4]: strings = rng.strftime('%Y-%m-%d %H:%M:%S').tolist()

In [5]: cProfile.run('to_datetime(strings)', sort='cumtime')
         273 function calls (270 primitive calls) in 0.017 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.017    0.017 <string>:1(<module>)
        1    0.001    0.001    0.017    0.017 datetimes.py:39(to_datetime)
        1    0.000    0.000    0.009    0.009 datetimes.py:207(_convert_listlike)
        1    0.008    0.008    0.008    0.008 {pandas._libs.tslib.array_to_datetime}
        1    0.000    0.000    0.007    0.007 datetimes.py:313(_maybe_convert_cache)
        1    0.000    0.000    0.007    0.007 algorithms.py:276(unique)
        1    0.005    0.005    0.005    0.005 {method 'unique' of 'pandas._libs.hashtable.PyObjectHashTable' objects}
        1    0.000    0.000    0.001    0.001 algorithms.py:164(_ensure_arraylike)
        2    0.001    0.001    0.001    0.001 {pandas._libs.lib.infer_dtype}
        1    0.000    0.000    0.001    0.001 algorithms.py:132(_reconstruct_data)
        2    0.001    0.000    0.001    0.000 {numpy.core.multiarray.array}
        1    0.000    0.000    0.001    0.001 algorithms.py:189(_get_hashtable_algo)
        1    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}
      114    0.000    0.000    0.000    0.000 {isinstance}
    11/10    0.000    0.000    0.000    0.000 common.py:1773(_get_dtype_type)
        1    0.000    0.000    0.000    0.000 {pandas._libs.lib.list_to_object_array}
      2/1    0.000    0.000    0.000    0.000 _decorators.py:86(wrapper)
        1    0.000    0.000    0.000    0.000 datetimes.py:269(__new__)
        3    0.000    0.000    0.000    0.000 common.py:1545(is_bool_dtype)
        1    0.000    0.000    0.000    0.000 algorithms.py:39(_ensure_data)
        6    0.000    0.000    0.000    0.000 dtypes.py:85(is_dtype)
        3    0.000    0.000    0.000    0.000 common.py:334(is_datetime64tz_dtype)
        3    0.000    0.000    0.000    0.000 common.py:297(is_datetime64_dtype)
        1    0.000    0.000    0.000    0.000 datetimes.py:579(_simple_new)
       12    0.000    0.000    0.000    0.000 generic.py:7(_check)
        3    0.000    0.000    0.000    0.000 common.py:478(is_categorical_dtype)
        1    0.000    0.000    0.000    0.000 common.py:1049(is_datetime64_ns_dtype)
        2    0.000    0.000    0.000    0.000 dtypes.py:428(construct_from_string)
        1    0.000    0.000    0.000    0.000 common.py:1722(_get_dtype)
        3    0.000    0.000    0.000    0.000 common.py:85(is_object_dtype)
       18    0.000    0.000    0.000    0.000 {getattr}
        2    0.000    0.000    0.000    0.000 common.py:409(is_period_dtype)
        3    0.000    0.000    0.000    0.000 abc.py:128(__instancecheck__)
        2    0.000    0.000    0.000    0.000 inference.py:234(is_list_like)
        2    0.000    0.000    0.000    0.000 dtypes.py:370(__new__)
        2    0.000    0.000    0.000    0.000 common.py:873(is_unsigned_integer_dtype)
        2    0.000    0.000    0.000    0.000 dtypes.py:554(is_dtype)
        2    0.000    0.000    0.000    0.000 common.py:824(is_signed_integer_dtype)
        2    0.000    0.000    0.000    0.000 common.py:1493(is_float_dtype)
        8    0.000    0.000    0.000    0.000 {hasattr}
       11    0.000    0.000    0.000    0.000 {issubclass}
      6/5    0.000    0.000    0.000    0.000 {len}
        1    0.000    0.000    0.000    0.000 common.py:442(is_interval_dtype)
        1    0.000    0.000    0.000    0.000 dtypes.py:676(is_dtype)
        4    0.000    0.000    0.000    0.000 _weakrefset.py:70(__contains__)
        5    0.000    0.000    0.000    0.000 {method 'startswith' of 'str' objects}
        1    0.000    0.000    0.000    0.000 numeric.py:463(asarray)
        2    0.000    0.000    0.000    0.000 dtypes.py:270(construct_from_string)
        2    0.000    0.000    0.000    0.000 {method 'search' of '_sre.SRE_Pattern' objects}
        4    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
        2    0.000    0.000    0.000    0.000 {pandas._libs.algos.ensure_object}
        1    0.000    0.000    0.000    0.000 base.py:558(__len__)
        1    0.000    0.000    0.000    0.000 base.py:552(_reset_identity)
        1    0.000    0.000    0.000    0.000 {built-in method __new__ of type object at 0x7f88fb67a4c0}
        1    0.000    0.000    0.000    0.000 {pandas._libs.tslibs.timezones.maybe_get_tz}
        1    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}
        1    0.000    0.000    0.000    0.000 base.py:467(_deepcopy_if_needed)
        1    0.000    0.000    0.000    0.000 frequencies.py:391(to_offset)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

jorisvandenbossche commented:
Given the potential slowdown, can't we add it as optional keyword? (so False by default)

jreback commented:
you need to benchmark on different sizes with and w/o duplicates to figure out the cutoff.

IOW n=[1000, 10000, 100000] or something

I suspect just making the cutoff higher will make ``cache=True`` a good defaultl.

mroeschke committed:
Use quicker unique check

mroeschke commented:
Apologies for the delay on my end with this PR.

It doesn't appear that there's a cutoff where a `cache=True` default avoids a slowdown, unless we know that all the dates are duplicates. Might be worth defaulting to `cache=False` with guidelines on when to set it to True. 

I've benchmarked the largest slowdown (parsing a date string) altering the number of dates to parse and different proportion of duplicate values (0 = all unique, 1 = all duplicate). The table below shows the timing ratios of `cache=True` time / `cache=False` times:
```
Proportion of Duplicate Values       0.0       0.2       0.4       0.6  \
Number of Values                                                         
1000.0                          1.717877  5.202867  4.594803  1.009279   
10000.0                         1.413974  3.235636  2.709920  2.395477   
100000.0                        1.878788  4.001020  3.401219  2.504916   
1000000.0                       2.320983  5.392797  3.579101  3.050970   

Proportion of Duplicate Values       0.8       1.0  
Number of Values                                    
1000.0                          1.009684  3.867605  
10000.0                         2.008572  1.289517  
100000.0                        1.798609  0.946278  
1000000.0                       2.138899  0.969332 
```

Code to generate table:
```
from timeit import Timer
from datetime import datetime, timedelta

import pandas as pd
import numpy as np
sizes = np.logspace(3, 6, 4)
proportions = np.linspace(0, 1, 6)
base = '1901-01-01 00:00:00'
fmt = '%Y-%m-%d %H:%M:%S'
unique_base = datetime.strptime(base, fmt) + timedelta(seconds=1)
results = []
for size in sizes:
    for prop_dupes in proportions:
        num_dupes = int(size * prop_dupes)
        num_uniques = int(size * (1 - prop_dupes))
        uniques = pd.date_range(unique_base, freq='s', periods=num_uniques).strftime(fmt).tolist()
        dupes = [base] * num_dupes
        data = dupes + uniques
        exec_time_true = Timer(lambda : pd.to_datetime(data, cache=True)).repeat(number=10)
        exec_time_false = Timer(lambda : pd.to_datetime(data, cache=False)).repeat(number=10)
        true_false_time_ratio = min(exec_time_true) / min(exec_time_false)
        result = [size, prop_dupes, true_false_time_ratio]
        results.append(result)
num_results = pd.DataFrame(results, columns=['Number of Values',
                                             'Proportion of Duplicate Values',
                                             'Cache True/False Time Ratio'])
num_results.pivot('Number of Values', 'Proportion of Duplicate Values', 'Cache True/False Time Ratio')
```

jorisvandenbossche commented:
@mroeschke do I understand that table correctly to interpret that also for the "all duplicate" case there is almost no speed-up ?

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+                result = Series(arg, name=name).map(convert_cache)
+                if isinstance(arg, Series):
+                    return result
+                elif box:
```
if u use _convert_list_like where u return None and do the boxing outside (iow where this is called) then you can make this simpler and avoid splitting logic 

now it’s ok 2 places

mroeschke replied:
Good call on return _convert_list_like instead of None.

However if I move the boxing outside this function, I think the code will be less DRY since I'll have to repeat the boxing logic twice for Indexes and list-like arguments. So essentially if:

```
def _maybe_convert_cache(...):
    if cache:
        ...
        return Series(arg, name=name).map(convert_cache)
    else: 
        return _convert_cache(...)
```
then the logic further down will be:

```
elif isinstance(arg, ABCIndexClass):
    result = _maybe_convert_cache(...)
    if box:
        return DatetimeIndex(...)
    else:
        return result.values
elif is_list_like(arg):
    result = _maybe_convert_cache(...)
    if box:
        return DatetimeIndex(...)
    else:
        return result.values
```

jreback replied:
see my comments above, this needs re-factoring.

mroeschke commented:
@jorisvandenbossche Correct (with maybe a slight speedup at ~100k duplicates), but keep in mind this is only for the case where the user is parsing ISO date strings. 

However, if a user is parsing a ISO date strings with tz offset (e.g `'1901-01-01 00:00:00-08:00'`) for example, the cache really helps (potentially ~2x speedup with just half duplicates):

```
Proportion of Duplicate Values       0.0       0.5       1.0
Number of Values                                            
1000.0                          1.013743  0.568563  0.054249
10000.0                         1.005625  0.518776  0.014114
100000.0                        1.008731  0.526137  0.010253
```

Bottom line, it appears that caching may improve performance depending on what the user is trying to parse, and the fact that it may introduce a slowdown is good reason to set `cache=False`. I can enhance the documentation to inform when `cache=True` should be used.

jreback commented:
>  can enhance the documentation to inform when cache=True should be used.

It possible we want a ``cache='infer'`` option (maybe as the default), where we do a small bit of introspection (e.g. parse first 1000 values or something) and 'guess' whether we should be using the cache things. But can do that in the future PR.

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             except (ValueError, TypeError):
                 raise e
 
+    def _maybe_convert_cache(arg, cache, box, format, name=None, tz=tz):
```
I would move this outside ``to_datetime``, making this function huge.

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+            If True, try to convert the dates with a cache
+            If False, short circuit and return None
+            Flag whether to cache the converted dates
+        box : boolean
```
boxing should be done outside

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
+        box : boolean
+            If True, return a DatetimeIndex  
+            if False, return an ndarray of values     
+        tz : String or None
```
same with these args. The wrapping should be done outside, ``_maybe_cache`` is a raw function that can always just return a ndarray. Then wrap it after its called. otherwise the logic is getting too cumbersome / duplicated.

mroeschke committed:
Move caching function outside to_datetime

mroeschke committed:
Pass most tests

mroeschke committed:
Skip test related to GH 18111, lint

mroeschke committed:
Update docstring

mroeschke commented:
I changed/updated the following:

1) Changed `cache=False` as the default after our prior discussion. Can change to 'infer' in a future PR after doing some inspection of what is being parsed.

2)  Created 2 functions outside of `to_datetime`, `_maybe_cache` which returns a cache array and `_convert_and_box_cache` which applies the cache to the data and boxes the result. 

`_maybe_cache` utilizes `_convert_listlike` for datetime conversion which is defined in `to_datetime` so I had to pass the function into `_maybe_cache`. Alternatively I could also move `_convert_listlike` out of `to_datetime`. Not sure which design is preferred. 

3) Skipped a test in `TestTimeConversionFormats.test_datetime_bool` when `cache=True` due to #18111

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -39,7 +39,7 @@ Deprecations
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
--
+- Added a keyword argument, `cache`, to :func:`to_datetime` that improved the performance of converting duplicate datetime arguments (:issue: `11665`)
```
let's move to 0.22.0

jschendel replied:
Additionally, there should be two backticks around <code>\`\`cache\`\`</code>, and the space between `:issue:` and the issue number should be removed.

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,39 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+
+def _convert_and_box_cache(arg, cache_array, box, errors, tz, name=None):
+    """Convert array of dates with a cache and box the result"""
+    from pandas import Series
```
you can import from pandas here directly

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,39 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+    result = Series(arg).map(cache_array)
+    if box:
+        if errors == 'ignore':
+            from pandas import Index
```
just import Index where you import others

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -111,7 +141,12 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
+    cache : boolean, default False
+        If False, use a cache of unique, converted dates to apply the datetime
+        conversion. May produce sigificant speed-up when parsing duplicate date
+        strings, especially ones with timezone offsets.
```
should be if True, change versionadded

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -369,15 +404,28 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
-        values = _convert_listlike(arg._values, True, format)
-        result = Series(values, index=arg.index, name=arg.name)
+        cache_array = _maybe_cache(arg, format, cache, tz, _convert_listlike)
+        if not cache_array.empty:
```
this still looks pretty duplicative, but I guess ok for now.

jreback reviewed (changes requested):
small comments. in follow up where we make ``cache='infer' `` maybe can rework the code a bit

mroeschke committed:
adjust imports, docs and move whatsnew

mroeschke committed:
Remove whitespace

mroeschke commented:
Adjusted the imports, fixed the doc comment, and moved the whatsnew note to v0.22.0. 

In a new issue we can highlight that `cache='infer'` should be added and maybe the cache conversion logic can be simplified.

jreback commented on asv_bench/benchmarks/timeseries.py:
```diff
@@ -381,6 +386,21 @@ def time_format_exact(self):
     def time_format_no_exact(self):
         to_datetime(self.s, format='%d%b%y', exact=False)
 
+    def time_cache_with_unique_seconds_and_unit(self):
```
i think these now need to matrix cache=True/False

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -70,7 +70,7 @@ Performance Improvements
 
 - Indexers on Series or DataFrame no longer create a reference cycle (:issue:`17956`)
--
+- Added a keyword argument, ``cache``, to :func:`to_datetime` that improved the performance of converting duplicate datetime arguments (:issue:`11665`)
```
say default is False

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,37 @@ def _guess_datetime_format_for_array(arr, **kwargs):
 
 
+def _maybe_cache(arg, format, cache, tz, _convert_listlike):
+    """Create a cache of unique dates from an array of dates"""
```
can u add Parameters and list what the args are

jreback replied:
and Returns

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,37 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+
+def _convert_and_box_cache(arg, cache_array, box, errors, tz, name=None):
+    """Convert array of dates with a cache and box the result"""
+    from pandas import Series, DatetimeIndex, Index
```
same

jreback commented on pandas/tests/indexes/datetimes/test_tools.py:
```diff
@@ -337,32 +364,38 @@ def test_to_datetime_tz_psycopg2(self):
 
-    def test_datetime_bool(self):
+    bool_skip = pytest.mark.skipif(True, reason="GH 18111")
+
```
just write this skip inline (don’t use a temp)
you can format so that it fits by putting things on separate lines

mroeschke committed:
Address comments

mroeschke committed:
Lint fix

mroeschke commented:
Added asvs with cache True/False matrix, expanded docs of helper functions, and adjusted decorator.

jreback commented on asv_bench/benchmarks/timeseries.py:
```diff
@@ -381,6 +386,36 @@ def time_format_exact(self):
     def time_format_no_exact(self):
         to_datetime(self.s, format='%d%b%y', exact=False)
 
+    def time_cache_true_with_unique_seconds_and_unit(self):
```
ok, I think there is a actually a way to do this with an asv matrix, but can ok for now

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,66 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+    Parameters
+    ----------
+    arg : integer, float, string, datetime, list, tuple, 1-d array of dates
+    format : string, strftime to parse time
```
can you format as a doc-string, IOW types on the first line, 2nd line for what it is doing.

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,66 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+    format : string, strftime to parse time
+    cache: boolean, whether to convert with cache
+    tz: string, timezone of the dates
+    _convert_listlike: function, conversion function to apply on dates
```
name this convert_listlike

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -36,9 +36,66 @@ def _guess_datetime_format_for_array(arr, **kwargs):
+    Parameters
+    ----------
+    arg : integer, float, string, datetime, list, tuple, 1-d array of dates
+    cache_array: Series, cache of converted, unique dates
```
same

jreback reviewed (changes requested):
some doc comments

mroeschke committed:
Move docs and adjust test

mroeschke committed:
Lint

jreback commented:
thanks @mroeschke nice PR!

look forward to ``cache='infer'`` !

pls open an issue for this.
